File: e2e_tests/trading/python/test_trading.py
#!/usr/bin/env python3
"""End-to-end authenticated-trading test for the Python SDK.

Mirrors `engine/sdk/tests/e2e_trading.rs` — every authenticated endpoint is
exercised against both Kalshi (BTC 15m) and Polymarket (BTC 5m) live markets
through the published Python `Exchange` wrapper.

Run from repo root:
    sdks/python/.venv/bin/python e2e_trading/python/test_trading.py

Output: writes a JSON results document to ../results/python.json and prints
a pass/skip/fail summary, exits non-zero if any case fails (skips don't
count toward failure — Polymarket auth-derive may be intermittent).
"""

from __future__ import annotations

import json
import math
import os
import sys
import time
import traceback
from pathlib import Path


# Load .env so credentials propagate to the native bindings.
def load_dotenv(path: Path) -> None:
    if not path.exists():
        return
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        os.environ.setdefault(k.strip(), v.strip())


REPO_ROOT = Path(__file__).resolve().parents[3]
load_dotenv(REPO_ROOT / ".env")
key_path = os.environ.get("KALSHI_PRIVATE_KEY_PATH", "")
if key_path and not os.path.isabs(key_path):
    os.environ["KALSHI_PRIVATE_KEY_PATH"] = str(REPO_ROOT / key_path)
RESULTS_DIR = REPO_ROOT / "e2e_tests" / "trading" / "results"
RESULTS_DIR.mkdir(parents=True, exist_ok=True)
RESULTS_PATH = RESULTS_DIR / "python.json"

from openpx import Exchange  # noqa: E402

KALSHI_CONFIG = {
    "api_key_id": os.environ.get("KALSHI_API_KEY_ID", ""),
    "private_key_path": os.environ.get("KALSHI_PRIVATE_KEY_PATH", ""),
}
POLY_CONFIG = {
    "private_key": os.environ.get("POLYMARKET_PRIVATE_KEY", ""),
    "funder": os.environ.get("POLYMARKET_FUNDER", ""),
    "signature_type": os.environ.get("POLYMARKET_SIGNATURE_TYPE", ""),
}


def make_exchange(eid: str):
    cfg = KALSHI_CONFIG if eid == "kalshi" else POLY_CONFIG
    cfg = {k: v for k, v in cfg.items() if v}
    return Exchange(eid, cfg)


# ---------------------------------------------------------------------------
# Result tracking + helpers
# ---------------------------------------------------------------------------

results: list[dict] = []


def _is_auth_unavailable(e: Exception) -> bool:
    msg = str(e).lower()
    return any(t in msg for t in (
        "could not derive api key", "cloudflare waf blocked",
        "cannot reach clob.polymarket.com", "l1 eip-712 signature",
        "no signing method", "private key required",
    ))


def _is_transient(e: Exception) -> bool:
    msg = str(e).lower()
    return any(t in msg for t in (
        "rate limit", "429", "timed out", "http error", "connection",
    ))


def case(name: str, expect: str = "ok"):
    def deco(fn):
        entry = {"name": name, "expect": expect, "status": "pending", "detail": None}
        try:
            fn(entry)
            if entry["status"] == "pending":
                if expect == "err":
                    entry["status"] = "fail"
                    entry["detail"] = "expected an error but call returned ok"
                else:
                    entry["status"] = "pass"
        except AssertionError as e:
            entry["status"] = "fail"
            entry["detail"] = f"assert: {e}\n{traceback.format_exc()}"
        except Exception as e:  # noqa: BLE001
            if expect == "err":
                entry["status"] = "pass"
                entry["detail"] = f"expected err: {e}"
            elif _is_transient(e):
                entry["status"] = "skip"
                entry["detail"] = f"transient: {e}"
            elif _is_auth_unavailable(e):
                entry["status"] = "skip"
                entry["detail"] = f"auth unavailable: {e}"
            else:
                entry["status"] = "fail"
                entry["detail"] = f"raised: {e}\n{traceback.format_exc()}"
        results.append(entry)
        symbol = {"pass": "✓", "fail": "✗", "skip": "·"}.get(entry["status"], "?")
        print(f"  {symbol} [{entry['status'].upper()}] {name}")
        if entry["status"] == "fail" and entry["detail"]:
            print(f"      → {str(entry['detail'])[:240]}")
        return fn

    return deco


def _unwrap(v):
    root = getattr(v, "root", None)
    if root is not None and not callable(root):
        return root
    return v


def get(d, key):
    if hasattr(d, key) and not isinstance(d, dict):
        return _unwrap(getattr(d, key))
    if isinstance(d, dict):
        return _unwrap(d.get(key))
    return _unwrap(getattr(d, key, None))


# ---------------------------------------------------------------------------
# Active BTC market discovery
# ---------------------------------------------------------------------------

def kalshi_btc_market(ex):
    """Resolve the currently-active KXBTC15M market — the one with the soonest
    close_time wins."""
    raw = ex.fetch_markets(status="active", series_ticker="KXBTC15M", limit=5)
    markets = raw.get("markets", []) if isinstance(raw, dict) else getattr(raw, "markets", [])
    if not markets:
        return None
    # Pick the soonest-closing one.
    def _close(m):
        t = m.get("close_time") if isinstance(m, dict) else getattr(m, "close_time", None)
        return str(t or "")
    return sorted(markets, key=_close)[0]


def polymarket_btc_market(ex):
    """Round Polymarket server time to nearest 5m boundary, find that event."""
    server = ex.fetch_server_time()
    # server is RFC3339 string from the wrapper.
    import datetime as dt
    parsed = dt.datetime.fromisoformat(server.replace("Z", "+00:00"))
    now_unix = int(parsed.timestamp())
    bucket = (now_unix // 300) * 300
    for offset in (0, 300, -300):
        b = bucket + offset
        et = f"btc-updown-5m-{b}"
        raw = ex.fetch_markets(status="active", event_ticker=et, limit=5)
        markets = raw.get("markets", []) if isinstance(raw, dict) else getattr(raw, "markets", [])
        if markets:
            return markets[0]
    return None


def _market_ticker(m):
    return m.get("ticker") if isinstance(m, dict) else getattr(m, "ticker", None)


def _first_token_id(m):
    outs = m.get("outcomes") if isinstance(m, dict) else getattr(m, "outcomes", None)
    if not outs:
        return None
    o = outs[0]
    return (o.get("token_id") if isinstance(o, dict) else getattr(o, "token_id", None))


def _best_bid(m):
    return m.get("best_bid") if isinstance(m, dict) else getattr(m, "best_bid", None)


def _tick_size(m):
    t = m.get("tick_size") if isinstance(m, dict) else getattr(m, "tick_size", None)
    return t if t else 0.01


def safe_resting_buy_price(best_bid, tick):
    candidate = (best_bid - tick * 5.0) if best_bid is not None else 0.05
    candidate = min(candidate, 0.05)
    snapped = math.floor(candidate / tick) * tick
    return max(tick, min(1.0 - tick, snapped))


def safe_size(eid: str, price: float) -> float:
    return 1.0 if eid == "kalshi" else math.ceil(5.5 / price)


# ---------------------------------------------------------------------------
# Per-exchange suites
# ---------------------------------------------------------------------------

def run_suite(eid: str, market_resolver):
    print(f"\n=== {eid} ===")
    try:
        ex = make_exchange(eid)
    except Exception as e:  # noqa: BLE001
        print(f"  · skipping {eid}: cannot construct exchange: {e}")
        return
    try:
        market = market_resolver(ex)
    except Exception as e:  # noqa: BLE001
        print(f"  · skipping {eid}: market discovery failed: {e}")
        return
    if not market:
        print(f"  · skipping {eid}: no active BTC market")
        return

    market_ticker = _market_ticker(market)
    asset_id = market_ticker if eid == "kalshi" else _first_token_id(market)
    tick = _tick_size(market)
    bid = _best_bid(market)
    price = safe_resting_buy_price(bid, tick)
    size = safe_size(eid, price)
    print(f"  · market={market_ticker} asset={asset_id} price={price} size={size} tick={tick}")

    # ----- Account -----
    @case(f"[{eid}] fetch_server_time")
    def _(entry):
        ts = ex.fetch_server_time()
        assert ts, "empty server time"

    @case(f"[{eid}] fetch_balance")
    def _(entry):
        bal = ex.fetch_balance()
        key = "USD" if eid == "kalshi" else "USDC"
        assert key in bal, f"missing {key} in {bal}"
        assert bal[key] >= 0.0, f"negative balance {bal}"

    @case(f"[{eid}] refresh_balance")
    def _(entry):
        ex.refresh_balance()  # returns None

    # ----- Positions -----
    @case(f"[{eid}] fetch_positions unfiltered")
    def _(entry):
        pos = ex.fetch_positions()
        for p in pos:
            assert get(p, "size") > 0.0, f"non-positive position {p}"

    @case(f"[{eid}] fetch_positions filtered")
    def _(entry):
        pos = ex.fetch_positions(market_ticker)
        for p in pos:
            assert get(p, "size") > 0.0, f"non-positive position {p}"

    # ----- Orders: lifecycle -----
    @case(f"[{eid}] order lifecycle: create→fetch→cancel")
    def _(entry):
        order = ex.create_order(asset_id, "yes", "buy", price, size, "gtc")
        order_id = get(order, "id")
        assert order_id, f"no id on placed order {order}"
        # Best-effort verify via fetch_order. Kalshi's GET /portfolio/orders/{id}
        # can briefly return "not found" right after create (indexing race).
        try:
            fetched = ex.fetch_order(order_id)
            assert get(fetched, "id") == order_id, "fetch_order id drift"
        except Exception as e:
            msg = str(e).lower()
            if "not found" in msg or "marketnotfound" in msg:
                entry["detail"] = f"fetch race: {e}; tolerated"
            elif not (_is_transient(e) or _is_auth_unavailable(e)):
                raise
        # And via fetch_open_orders unfiltered
        try:
            opens = ex.fetch_open_orders()
            ids = [get(o, "id") for o in opens]
            if order_id not in ids and get(order, "status") not in ("filled", "Filled"):
                # not necessarily a fail — race window with fills/cancels
                entry["detail"] = f"open list missed our id {order_id}; tolerated"
        except Exception as e:
            if not (_is_transient(e) or _is_auth_unavailable(e)):
                raise
        # Always try to cancel
        try:
            cancelled = ex.cancel_order(order_id)
            assert get(cancelled, "status") in ("cancelled", "Cancelled"), \
                f"cancel returned non-cancelled status {cancelled}"
        except Exception as e:
            msg = str(e).lower()
            if "not found" in msg or "filled" in msg:
                pass  # already filled or gone — acceptable
            elif _is_transient(e) or _is_auth_unavailable(e):
                pass
            else:
                raise

    # ----- Orders: adversarial -----
    @case(f"[{eid}] create_order zero price", expect="err")
    def _(entry):
        ex.create_order(asset_id, "yes", "buy", 0.0, size, "gtc")

    @case(f"[{eid}] create_order one price", expect="err")
    def _(entry):
        ex.create_order(asset_id, "yes", "buy", 1.0, size, "gtc")

    @case(f"[{eid}] create_order negative size", expect="err")
    def _(entry):
        ex.create_order(asset_id, "yes", "buy", price, -size, "gtc")

    # ----- fetch_order / cancel_order with bogus id -----
    fake_id = "00000000-0000-0000-0000-000000000000" if eid == "kalshi" \
        else "0xdeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddead"

    @case(f"[{eid}] fetch_order unknown id", expect="err")
    def _(entry):
        ex.fetch_order(fake_id)

    @case(f"[{eid}] cancel_order unknown id", expect="err")
    def _(entry):
        ex.cancel_order(fake_id)

    # ----- Batch: empty / oversize -----
    @case(f"[{eid}] create_orders_batch empty")
    def _(entry):
        out = ex.create_orders_batch([])
        assert isinstance(out, list) and len(out) == 0, f"empty batch returned {out}"

    @case(f"[{eid}] create_orders_batch oversize", expect="err")
    def _(entry):
        cap = 21 if eid == "kalshi" else 16
        reqs = [
            {"asset_id": asset_id, "outcome": "yes", "side": "buy",
             "price": 0.05, "size": size, "order_type": "gtc"}
            for _ in range(cap)
        ]
        # Polymarket has hard cap 15; Kalshi may accept then we cancel.
        out = ex.create_orders_batch(reqs)
        # Cleanup if it accepted.
        for o in out or []:
            try:
                ex.cancel_order(get(o, "id"))
            except Exception:
                pass
        if eid == "polymarket":
            raise AssertionError(f"oversize batch ({cap}) was accepted")

    # ----- Bulk cancels -----
    @case(f"[{eid}] cancel_all_orders unfiltered")
    def _(entry):
        out = ex.cancel_all_orders()
        assert isinstance(out, list)

    @case(f"[{eid}] cancel_all_orders filtered")
    def _(entry):
        out = ex.cancel_all_orders(asset_id)
        assert isinstance(out, list)

    # ----- Open orders -----
    @case(f"[{eid}] fetch_open_orders unfiltered")
    def _(entry):
        out = ex.fetch_open_orders()
        for o in out:
            assert get(o, "id"), f"order missing id: {o}"

    @case(f"[{eid}] fetch_open_orders filtered")
    def _(entry):
        out = ex.fetch_open_orders(asset_id)
        assert isinstance(out, list)

    # ----- Fills -----
    @case(f"[{eid}] fetch_fills unfiltered limit=10")
    def _(entry):
        fills = ex.fetch_fills(limit=10)
        assert len(fills) <= 10, f"limit not honored: {len(fills)}"
        for f in fills:
            p = get(f, "price")
            s = get(f, "size")
            assert s > 0.0, f"non-positive fill size {f}"
            assert 0.0 < p < 1.0, f"fill price out of range {f}"

    @case(f"[{eid}] fetch_fills filtered limit=5")
    def _(entry):
        fills = ex.fetch_fills(market_ticker=market_ticker, limit=5)
        assert len(fills) <= 5

    # ----- Trades (public) -----
    @case(f"[{eid}] fetch_trades basic")
    def _(entry):
        trades = ex.fetch_trades(market_ticker, limit=20)
        # The wrapper returns dict with 'trades' key from native binding.
        items = trades.get("trades", []) if isinstance(trades, dict) else trades
        for t in items:
            p = get(t, "price")
            s = get(t, "size")
            assert s > 0.0, f"trade size non-positive {t}"
            assert 0.0 < p < 1.0, f"trade price out of (0,1) {t}"

    @case(f"[{eid}] fetch_trades with time window")
    def _(entry):
        now = int(time.time())
        trades = ex.fetch_trades(market_ticker, start_ts=now - 3600, end_ts=now, limit=50)
        items = trades.get("trades", []) if isinstance(trades, dict) else trades
        # Loose bound: trades within window or a small skew tolerance.
        assert isinstance(items, list)


# ---------------------------------------------------------------------------
# Drive both exchanges
# ---------------------------------------------------------------------------

def main() -> int:
    run_suite("kalshi", kalshi_btc_market)
    run_suite("polymarket", polymarket_btc_market)

    pass_n = sum(1 for r in results if r["status"] == "pass")
    skip_n = sum(1 for r in results if r["status"] == "skip")
    fail_n = sum(1 for r in results if r["status"] == "fail")

    summary = {
        "total": len(results),
        "pass": pass_n,
        "skip": skip_n,
        "fail": fail_n,
        "results": results,
    }
    RESULTS_PATH.write_text(json.dumps(summary, indent=2, default=str))
    print(f"\nPASS {pass_n}\nSKIP {skip_n}\nFAIL {fail_n}")
    if fail_n > 0:
        print("\nFailing cases:")
        for r in results:
            if r["status"] == "fail":
                print(f"  - {r['name']}: {str(r['detail'])[:200]}")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: e2e_trading/typescript/test_trading.mjs
// End-to-end authenticated-trading test for the TypeScript / NAPI SDK.
//
// Mirrors `engine/sdk/tests/e2e_trading.rs` and the Python suite — every
// authenticated endpoint is exercised against both Kalshi (BTC 15m) and
// Polymarket (BTC 5m) live markets through the published Node addon.
//
// Run from repo root:
//     node e2e_trading/typescript/test_trading.mjs
//
// Output: writes ../results/typescript.json + pass/skip/fail summary to
// stdout, exits non-zero if any case fails.

import { Exchange } from '../../sdks/typescript/index.js';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const REPO_ROOT = path.resolve(__dirname, '../..');
const RESULTS_DIR = path.join(REPO_ROOT, 'e2e_trading', 'results');
fs.mkdirSync(RESULTS_DIR, { recursive: true });
const RESULTS_PATH = path.join(RESULTS_DIR, 'typescript.json');

function loadDotenv(p) {
    if (!fs.existsSync(p)) return;
    for (const line of fs.readFileSync(p, 'utf8').split('\n')) {
        const t = line.trim();
        if (!t || t.startsWith('#') || !t.includes('=')) continue;
        const [k, ...rest] = t.split('=');
        const v = rest.join('=');
        if (process.env[k.trim()] === undefined) process.env[k.trim()] = v.trim();
    }
}
loadDotenv(path.join(REPO_ROOT, '.env'));

const keyPath = process.env.KALSHI_PRIVATE_KEY_PATH;
if (keyPath && !path.isAbsolute(keyPath)) {
    process.env.KALSHI_PRIVATE_KEY_PATH = path.join(REPO_ROOT, keyPath);
}

function makeExchange(eid) {
    const cfg = eid === 'kalshi'
        ? {
            api_key_id: process.env.KALSHI_API_KEY_ID,
            private_key_path: process.env.KALSHI_PRIVATE_KEY_PATH,
        }
        : {
            private_key: process.env.POLYMARKET_PRIVATE_KEY,
            funder: process.env.POLYMARKET_FUNDER,
            signature_type: process.env.POLYMARKET_SIGNATURE_TYPE,
        };
    for (const k of Object.keys(cfg)) if (!cfg[k]) delete cfg[k];
    return new Exchange(eid, cfg);
}

// ---------------------------------------------------------------------------
// Result tracking + helpers
// ---------------------------------------------------------------------------

const results = [];

function isAuthUnavailable(e) {
    const m = (e?.message || String(e)).toLowerCase();
    return m.includes('could not derive api key')
        || m.includes('cloudflare waf blocked')
        || m.includes('cannot reach clob.polymarket.com')
        || m.includes('l1 eip-712 signature')
        || m.includes('no signing method')
        || m.includes('private key required');
}

function isTransient(e) {
    const m = (e?.message || String(e)).toLowerCase();
    return m.includes('rate limit') || m.includes('429') || m.includes('timed out')
        || m.includes('http error') || m.includes('connection');
}

async function runCase(name, expect, body) {
    const entry = { name, expect, status: 'pending', detail: null };
    try {
        await body(entry);
        if (entry.status === 'pending') {
            if (expect === 'err') {
                entry.status = 'fail';
                entry.detail = 'expected an error but call returned ok';
            } else {
                entry.status = 'pass';
            }
        }
    } catch (e) {
        if (expect === 'err') {
            entry.status = 'pass';
            entry.detail = `expected err: ${e.message || e}`;
        } else if (isTransient(e)) {
            entry.status = 'skip';
            entry.detail = `transient: ${e.message || e}`;
        } else if (isAuthUnavailable(e)) {
            entry.status = 'skip';
            entry.detail = `auth unavailable: ${e.message || e}`;
        } else {
            entry.status = 'fail';
            entry.detail = `raised: ${e.message || e}\n${e.stack || ''}`;
        }
    }
    results.push(entry);
    const sym = { pass: '✓', fail: '✗', skip: '·' }[entry.status] || '?';
    console.log(`  ${sym} [${entry.status.toUpperCase()}] ${name}`);
    if (entry.status === 'fail' && entry.detail) {
        console.log(`      → ${String(entry.detail).slice(0, 240)}`);
    }
}

// ---------------------------------------------------------------------------
// BTC market discovery (per exchange)
// ---------------------------------------------------------------------------

async function kalshiBtcMarket(ex) {
    const raw = await ex.fetchMarkets('active', null, ['KXBTC15M'].length ? null : null,
        'KXBTC15M', null, 5);
    const markets = raw?.markets || [];
    if (!markets.length) return null;
    return markets.slice().sort((a, b) => String(a.close_time || '').localeCompare(String(b.close_time || '')))[0];
}

async function polymarketBtcMarket(ex) {
    const isoTs = await ex.fetchServerTime();
    const nowUnix = Math.floor(new Date(isoTs).getTime() / 1000);
    const bucket = Math.floor(nowUnix / 300) * 300;
    for (const offset of [0, 300, -300]) {
        const b = bucket + offset;
        const et = `btc-updown-5m-${b}`;
        const raw = await ex.fetchMarkets('active', null, null, null, et, 5);
        const markets = raw?.markets || [];
        if (markets.length) return markets[0];
    }
    return null;
}

function safeRestingBuyPrice(bestBid, tick) {
    const t = tick || 0.01;
    let candidate = bestBid != null ? bestBid - 5 * t : 0.05;
    candidate = Math.min(candidate, 0.05);
    const snapped = Math.floor(candidate / t) * t;
    return Math.max(t, Math.min(1.0 - t, snapped));
}

function safeSize(eid, price) {
    return eid === 'kalshi' ? 1.0 : Math.ceil(5.5 / price);
}

// ---------------------------------------------------------------------------
// Per-exchange suite
// ---------------------------------------------------------------------------

async function runSuite(eid, marketResolver) {
    console.log(`\n=== ${eid} ===`);
    let ex;
    try {
        ex = makeExchange(eid);
    } catch (e) {
        console.log(`  · skipping ${eid}: cannot construct: ${e.message || e}`);
        return;
    }

    let market;
    try {
        market = await marketResolver(ex);
    } catch (e) {
        console.log(`  · skipping ${eid}: market discovery failed: ${e.message || e}`);
        return;
    }
    if (!market) {
        console.log(`  · skipping ${eid}: no active BTC market`);
        return;
    }

    const marketTicker = market.ticker;
    const assetId = eid === 'kalshi' ? marketTicker : market.outcomes?.[0]?.token_id;
    const tick = market.tick_size || 0.01;
    const price = safeRestingBuyPrice(market.best_bid, tick);
    const size = safeSize(eid, price);
    console.log(`  · market=${marketTicker} asset=${assetId} price=${price} size=${size} tick=${tick}`);

    // Account
    await runCase(`[${eid}] fetchServerTime`, 'ok', async () => {
        const ts = await ex.fetchServerTime();
        if (!ts) throw new Error('empty server time');
    });

    await runCase(`[${eid}] fetchBalance`, 'ok', async () => {
        const bal = await ex.fetchBalance();
        const key = eid === 'kalshi' ? 'USD' : 'USDC';
        if (!(key in bal)) throw new Error(`missing ${key} in ${JSON.stringify(bal)}`);
        if (bal[key] < 0) throw new Error(`negative balance ${bal[key]}`);
    });

    await runCase(`[${eid}] refreshBalance`, 'ok', async () => {
        await ex.refreshBalance();
    });

    // Positions
    await runCase(`[${eid}] fetchPositions unfiltered`, 'ok', async () => {
        const pos = await ex.fetchPositions();
        for (const p of pos) {
            if (!(p.size > 0)) throw new Error(`non-positive position ${JSON.stringify(p)}`);
        }
    });

    await runCase(`[${eid}] fetchPositions filtered`, 'ok', async () => {
        const pos = await ex.fetchPositions(marketTicker);
        for (const p of pos) {
            if (!(p.size > 0)) throw new Error(`non-positive position ${JSON.stringify(p)}`);
        }
    });

    // Order lifecycle
    await runCase(`[${eid}] order lifecycle: create→fetch→cancel`, 'ok', async () => {
        const order = await ex.createOrder(assetId, 'yes', 'buy', price, size, 'gtc');
        const orderId = order?.id;
        if (!orderId) throw new Error(`no id on placed order ${JSON.stringify(order)}`);
        // Kalshi's GET /portfolio/orders/{id} can briefly return "not found"
        // right after create (indexing race) — tolerate it as a soft-fail.
        try {
            const fetched = await ex.fetchOrder(orderId);
            if (fetched.id !== orderId) throw new Error('fetch_order id drift');
        } catch (e) {
            const m = (e.message || String(e)).toLowerCase();
            if (m.includes('not found') || m.includes('marketnotfound')) {
                entry.detail = `fetch race: ${e.message}; tolerated`;
            } else if (!(isTransient(e) || isAuthUnavailable(e))) {
                throw e;
            }
        }
        try {
            const cancelled = await ex.cancelOrder(orderId);
            const status = (cancelled.status || '').toLowerCase();
            if (status !== 'cancelled') throw new Error(`cancel returned non-cancelled ${JSON.stringify(cancelled)}`);
        } catch (e) {
            const m = (e.message || String(e)).toLowerCase();
            if (m.includes('not found') || m.includes('filled')) return;
            if (isTransient(e) || isAuthUnavailable(e)) return;
            throw e;
        }
    });

    // Adversarial create
    await runCase(`[${eid}] createOrder zero price`, 'err', async () => {
        await ex.createOrder(assetId, 'yes', 'buy', 0.0, size, 'gtc');
    });

    await runCase(`[${eid}] createOrder one price`, 'err', async () => {
        await ex.createOrder(assetId, 'yes', 'buy', 1.0, size, 'gtc');
    });

    await runCase(`[${eid}] createOrder negative size`, 'err', async () => {
        await ex.createOrder(assetId, 'yes', 'buy', price, -size, 'gtc');
    });

    // Fake order id
    const fakeId = eid === 'kalshi'
        ? '00000000-0000-0000-0000-000000000000'
        : '0xdeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddead';

    await runCase(`[${eid}] fetchOrder unknown id`, 'err', async () => {
        await ex.fetchOrder(fakeId);
    });

    await runCase(`[${eid}] cancelOrder unknown id`, 'err', async () => {
        await ex.cancelOrder(fakeId);
    });

    // Batch
    await runCase(`[${eid}] createOrdersBatch empty`, 'ok', async () => {
        const out = await ex.createOrdersBatch([]);
        if (!Array.isArray(out) || out.length !== 0) {
            throw new Error(`empty batch returned ${JSON.stringify(out)}`);
        }
    });

    await runCase(`[${eid}] createOrdersBatch oversize`, 'err', async () => {
        const cap = eid === 'kalshi' ? 21 : 16;
        const reqs = [];
        for (let i = 0; i < cap; i++) {
            reqs.push({ asset_id: assetId, outcome: 'yes', side: 'buy', price: 0.05, size, order_type: 'gtc' });
        }
        const out = await ex.createOrdersBatch(reqs);
        for (const o of out || []) {
            try { await ex.cancelOrder(o.id); } catch {}
        }
        if (eid === 'polymarket') {
            throw new Error(`oversize batch (${cap}) was accepted`);
        }
    });

    await runCase(`[${eid}] cancelAllOrders unfiltered`, 'ok', async () => {
        const out = await ex.cancelAllOrders();
        if (!Array.isArray(out)) throw new Error(`expected array, got ${typeof out}`);
    });

    await runCase(`[${eid}] cancelAllOrders filtered`, 'ok', async () => {
        const out = await ex.cancelAllOrders(assetId);
        if (!Array.isArray(out)) throw new Error(`expected array, got ${typeof out}`);
    });

    await runCase(`[${eid}] fetchOpenOrders unfiltered`, 'ok', async () => {
        const out = await ex.fetchOpenOrders();
        for (const o of out) if (!o.id) throw new Error(`order missing id ${JSON.stringify(o)}`);
    });

    await runCase(`[${eid}] fetchOpenOrders filtered`, 'ok', async () => {
        const out = await ex.fetchOpenOrders(assetId);
        if (!Array.isArray(out)) throw new Error(`expected array, got ${typeof out}`);
    });

    await runCase(`[${eid}] fetchFills unfiltered limit=10`, 'ok', async () => {
        const fills = await ex.fetchFills(undefined, 10);
        if (fills.length > 10) throw new Error(`limit not honored: ${fills.length}`);
        for (const f of fills) {
            if (!(f.size > 0)) throw new Error(`non-positive fill size ${JSON.stringify(f)}`);
            if (!(f.price > 0 && f.price < 1)) throw new Error(`fill price out of range ${JSON.stringify(f)}`);
        }
    });

    await runCase(`[${eid}] fetchFills filtered limit=5`, 'ok', async () => {
        const fills = await ex.fetchFills(marketTicker, 5);
        if (fills.length > 5) throw new Error(`limit not honored: ${fills.length}`);
    });

    await runCase(`[${eid}] fetchTrades basic`, 'ok', async () => {
        const trades = await ex.fetchTrades(marketTicker, null, null, 20);
        const items = trades?.trades || [];
        for (const t of items) {
            if (!(t.size > 0)) throw new Error(`trade size non-positive ${JSON.stringify(t)}`);
            if (!(t.price > 0 && t.price < 1)) throw new Error(`trade price out of (0,1) ${JSON.stringify(t)}`);
        }
    });

    await runCase(`[${eid}] fetchTrades with time window`, 'ok', async () => {
        const now = Math.floor(Date.now() / 1000);
        const trades = await ex.fetchTrades(marketTicker, now - 3600, now, 50);
        const items = trades?.trades || [];
        if (!Array.isArray(items)) throw new Error('trades not iterable');
    });
}

// ---------------------------------------------------------------------------
// Drive
// ---------------------------------------------------------------------------

async function main() {
    await runSuite('kalshi', kalshiBtcMarket);
    await runSuite('polymarket', polymarketBtcMarket);

    const passN = results.filter(r => r.status === 'pass').length;
    const skipN = results.filter(r => r.status === 'skip').length;
    const failN = results.filter(r => r.status === 'fail').length;

    fs.writeFileSync(RESULTS_PATH, JSON.stringify({
        total: results.length, pass: passN, skip: skipN, fail: failN, results,
    }, null, 2));

    console.log(`\nPASS ${passN}\nSKIP ${skipN}\nFAIL ${failN}`);
    if (failN > 0) {
        console.log('\nFailing cases:');
        for (const r of results) {
            if (r.status === 'fail') console.log(`  - ${r.name}: ${String(r.detail).slice(0, 200)}`);
        }
        process.exit(1);
    }
}

main().catch(e => {
    console.error('runner crashed:', e);
    process.exit(2);
});
